File: biddo_web/src/core/repositories/account.ts
import { RequestMaker, RequestType } from '../services/request-maker'
import { Account } from '../domain/account'

class AccountsRepository {
  async loadLoggedIn(): Promise<Account | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: '/account',
      })) as Record<string, unknown>
      return Account.fromJSON(response)
    } catch (error) {
      console.error('Error loading logged-in user:', error)
      return null
    }
  }

  async delete(): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.DELETE,
        path: '/account',
      })
      return true
    } catch (error) {
      console.error('Error deleting account:', error)
      return false
    }
  }

  async loadAccountDetails(accountId: string): Promise<Account | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: `/account/details/${accountId}`,
      })) as Record<string, unknown>
      return Account.fromJSON(response)
    } catch (error) {
      console.error('Error loading account details:', error)
      return null
    }
  }

  async askForVerification() {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.PUT,
        path: '/account/requestVerification',
      })
      return true
    } catch (error) {
      console.error('Error asking for verification:', error)
      return false
    }
  }

  async update(account: Account, asset?: File): Promise<Account | null> {
    const {
      name,
      allowedNotifications,
      acceptedTermsAndCondition,
      introDone,
      introSkipped,
      categoriesSetupDone,
      preferredCategoriesIds,
      locationLatLng,
      locationPretty,
      selectedCurrencyId,
    } = account

    const payload = {
      meta: account.meta?.asObject(),
      name,
      allowedNotifications,
      categoriesSetupDone,
      locationLatLng: JSON.stringify([locationLatLng?.lat ?? null, locationLatLng?.lng ?? null]),
      locationPretty,
      selectedCurrencyId,
      acceptedTermsAndCondition,
      introDone,
      introSkipped,
      preferredCategoriesIds,
    } as Record<string, unknown>

    const form = new FormData()
    if (asset) {
      Object.keys(payload).forEach((key) => {
        const value = payload[key]
        form.append(key, typeof value === 'string' ? value : JSON.stringify(value))
      })

      form.append('files', asset)
    }
    const finalPayload = asset ? form : JSON.stringify(payload)

    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: '/account',
        payload: finalPayload,
        contentType: asset ? 'multipart/form-data' : 'application/json',
      })) as Record<string, unknown>
      return Account.fromJSON(response)
    } catch (error) {
      console.error('Could not update account:', error)
      return null
    }
  }

  async search(keyword: string, page: number = 0): Promise<Account[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: '/account/search',
        payload: JSON.stringify({ keyword, page }),
        contentType: 'application/json',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Account.fromJSON(el))
    } catch (error) {
      console.error('Error searching accounts:', error)
      return []
    }
  }

  async loadStats(): Promise<Record<string, number> | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: '/account/stats',
      })) as Record<string, number>
      return response
    } catch (error) {
      console.error('Error loading account stats:', error)
      return null
    }
  }

  async blockAccount(accountId: string): Promise<void> {
    await RequestMaker.makeRequest({
      method: RequestType.PUT,
      path: `/account/update/blocked/block/${accountId}`,
    })
  }

  async unblockAccount(accountId: string): Promise<void> {
    await RequestMaker.makeRequest({
      method: RequestType.PUT,
      path: `/account/update/blocked/unblock/${accountId}`,
    })
  }
}

const accountsRepository = new AccountsRepository()
export { accountsRepository as AccountsRepository }

File: biddo_client/lib/core/repositories/account.dart
import 'dart:convert';

import 'package:get/get.dart';
import 'package:image_picker/image_picker.dart';
import 'package:dio/dio.dart' as dio_lib;

import '../controllers/image_picker.dart';
import '../models/account.dart';
import '../services/image_compress.dart';
import 'base.dart';

class AccountsRepository {
  final dio = Get.find<Api>();
  final imageCompressService = Get.find<ImageCompressService>();

  Future<Account?> loadLoggedIn() async {
    try {
      var response = await dio.api.get('/account');
      return Account.fromJSON(response.data);
    } catch (error, stacktrace) {
      print('error loading logged in user: $error, $stacktrace');
      return null;
    }
  }

  Future<bool> askForVerification() async {
    try {
      await dio.api.put('/account/requestVerification');
      return true;
    } catch (error) {
      print('error asking for verification: $error');
      return false;
    }
  }

  Future<bool> delete(Account account) async {
    try {
      await dio.api.delete('/account');
      return true;
    } catch (error, stacktrace) {
      print('error deleting account: $error, $stacktrace');
      return false;
    }
  }

  Future<Account?> loadAccountDetails(String accountId) async {
    try {
      var response = await dio.api.get('/account/details/$accountId');
      return Account.fromJSON(response.data);
    } catch (error) {
      print('error loading account details: $error');
      return null;
    }
  }

  Future<Account?> update(
    Account account, [
    AssetEntityPath? galleryAsset,
    XFile? cameraAsset,
  ]) async {
    List<dio_lib.MultipartFile> files = [];

    if (galleryAsset != null) {
      var compressed = await imageCompressService.compressFileByPath(
          galleryAsset.path, 250, 250);

      files.add(dio_lib.MultipartFile.fromFileSync(
        compressed.path,
        filename: galleryAsset.name,
      ));
    }

    if (cameraAsset != null) {
      var compressed = await imageCompressService.compressFile(cameraAsset);

      files.add(dio_lib.MultipartFile.fromFileSync(
        compressed.path,
        filename: cameraAsset.name,
      ));
    }

    try {
      var data = dio_lib.FormData.fromMap({
        'name': account.name,
        'deviceFCMToken': account.deviceFCMToken,
        'meta': account.meta?.asObject(),
        'allowedNotifications': account.allowedNotifications?.asObject(),
        'acceptedTermsAndCondition': account.acceptedTermsAndCondition,
        'introDone': account.introDone,
        'introSkipped': account.introSkipped,
        'categoriesSetupDone': account.categoriesSetupDone,
        'preferredCategoriesIds': account.preferredCategoriesIds,
        'files': files,
        'selectedCurrencyId': account.selectedCurrencyId,
        'locationLatLng': jsonEncode(account.locationLatLng),
        'locationPretty': account.locationPretty,
      });

      var response = await dio.api.post('/account', data: data);
      return Account.fromJSON(response.data);
    } catch (error) {
      print('could not update account: $error');
      return null;
    }
  }

  Future<List<Account>> search(String keyword, [int page = 0]) async {
    try {
      var response = await dio.api.post('/account/search', data: {
        'keyword': keyword,
        'page': page,
      });

      return List<Account>.from(
        response.data.map(
          (el) => Account.fromJSON(el),
        ),
      );
    } catch (error) {
      print('Error searching accounts: $error');
      return [];
    }
  }

  Future<dynamic> loadStats() async {
    try {
      var response = await dio.api.get('/account/stats');
      return response.data;
    } catch (error) {
      print('Error loading account stats: $error');
      return null;
    }
  }

  Future<void> blockAccount(String accountId) {
    return dio.api.put('/account/update/blocked/block/$accountId');
  }

  Future<void> unblockAccount(String accountId) {
    return dio.api.put('/account/update/blocked/unblock/$accountId');
  }
}
